dl/translation: streamline error handling

bharathv opened the pull request:
Cleaned up error handling further to totally rely on error code propagation from the writers all the way to translation.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes
* none

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

bharathv committed:
dl/translation: streamline error handling

Removes exception based error handling by propagating error codes
from each translation iteration.

Without this change the error handling was mix of eerror codes and
exception catching.

bharathv commented:
/dt

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--c92d7a40-1126-4f37-960e-c3a607d636aa-->
<!--TEST RESULTS-->
<!--94706337-a107-4f2f-8b8f-730510608bea-->
<!--TEST RESULTS-->
<!--ce496a11-505a-48ef-845c-c8f75f4284b4-->
<!--TEST RESULTS-->
<!--9b286af5-ed61-47ef-b917-6fe1bf9a66aa-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63843">build#63843</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/63843#0195db1c-7f81-4022-b380-687b85ca726b | FLAKY | 1/2 |
| rptest.tests.node_pool_migration_test.NodePoolMigrationTest.test_migrating_redpanda_nodes_to_new_pool.balancing_mode=node_add.test_mode=TestMode.FAST_MOVES.cleanup_policy=compact.delete | ducktape | https://buildkite.com/redpanda/redpanda/builds/63843#0195db09-58e8-4e84-aa72-c093587100c8 | FLAKY | 1/2 |
| rptest.tests.partition_force_reconfiguration_test.PartitionForceReconfigurationTest.test_basic_reconfiguration.acks=-1.restart=False.controller_snapshots=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/63843#0195db1c-7f7f-4ea3-945e-b03f9159e924 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64344">build#64344</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.True | ducktape | https://buildkite.com/redpanda/redpanda/builds/64344#01961d2e-4ff8-406d-bfb0-060c9e59586d | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/64344#01961d2e-4ff8-4084-b990-69ffbf6669c4 | FLAKY | 1/2 |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/64344#01961d2b-3bab-452b-ae7c-82df696428b0 | FLAKY | 1/4 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64877">build#64877</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_raft_rpunit.gtest_raft_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/64877#01965ea9-1851-47c1-a431-55e5676c3ddd | FLAKY | 1/2 |
| rptest.tests.datalake.disk_budget_test.DatalakeDiskUsageTest.test_idle_finish.num_partitions=10.concurrent_translations=4.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/64877#01965f04-1d9f-480e-8826-27dea3734ff5 | FAIL | 0/1 |
| rptest.tests.partition_force_reconfiguration_test.PartitionForceReconfigurationTest.test_basic_reconfiguration.acks=-1.restart=True.controller_snapshots=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/64877#01965eff-7653-488c-8362-f8b54594dfef | FLAKY | 20/21 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=False.with_iceberg=True.with_chunked_compaction=False.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/64877#01965f04-1da1-4503-a232-557cb9d067b7 | FLAKY | 19/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65159">build#65159</a></summary>


| test_id                                                                                                                                                                                                                        | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_conflicting_names                                                                                                                                             | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65159#01967ae6-7e9d-4aa7-8eb7-00473e5a903f | FLAKY         | 3/21     |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_conflicting_names                                                                                                                                             | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65159#01967b02-ac1a-4224-89fb-f1552a35388c | FLAKY         | 1/8      |
| rptest.tests.datalake.custom_partitioning_test.DatalakeCustomPartitioningTest.test_many_partitions.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65159#01967b02-ac1a-4224-89fb-f1552a35388c | FAIL          | 0/21     |
| rptest.tests.datalake.disk_budget_test.DatalakeDiskUsageTest.test_idle_finish.num_partitions=10.concurrent_translations=4.cloud_storage_type=CloudStorageType.S3                                                               | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65159#01967b02-ac19-4add-91bc-43c68e0ce9a3 | FAIL          | 0/21     |
| rptest.tests.datalake.disk_budget_test.DatalakeDiskUsageTest.test_idle_finish.num_partitions=40.concurrent_translations=4.cloud_storage_type=CloudStorageType.S3                                                               | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65159#01967b02-ac1a-4224-89fb-f1552a35388c | FAIL          | 0/1      |
| rptest.tests.datalake.translation_interruption_test.DatalakeTranslationInterruptionsTest.test_oom_interruptions.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.catalog_type=CatalogType.REST_HADOOP | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65159#01967ae6-7e9a-40d6-b2ab-4b1db554b236 | FAIL          | 0/21     |
| rptest.tests.datalake.translation_interruption_test.DatalakeTranslationInterruptionsTest.test_oom_interruptions.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.catalog_type=CatalogType.REST_HADOOP | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65159#01967b02-ac19-4add-91bc-43c68e0ce9a3 | FAIL          | 0/21     |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic                                                                                                                                                | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65159#01967b02-ac1b-4a7e-ac69-89fee9852dab | FLAKY         | 15/21    |
| rptest.tests.simple_e2e_test.SimpleEndToEndTest.test_consumer_interruption                                                                                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65159#01967b02-ac19-4add-91bc-43c68e0ce9a3 | FLAKY         | 10/21    |
| rptest.tests.tiered_storage_model_test.TieredStorageTest.test_tiered_storage.cloud_storage_type_and_url_style=.CloudStorageType.S3.1.path.test_case=.TS_Read==True.TS_TxRangeMaterialized==True                                | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65159#01967ae6-7e9d-4aa7-8eb7-00473e5a903f | FAIL          | 0/1      |
| rptest.transactions.producers_api_test.ProducersAdminAPITest.test_producers_state_api_during_load                                                                                                                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65159#01967ae6-7e9d-4aa7-8eb7-00473e5a903f | FLAKY         | 20/21    |
</details>
<!--/body-->

bharathv commented:
/dt

bharathv committed:
dl/translation: move record_mux/translation_task to translation/

No logical changes.

Moves translation related stuff like record multipler and
translation_task into src/v/datalake/translation/[tests] for better
dependency hierarchy. Needed for a future commit that adds certain
exceptions specific to translation and it is a bit odd to have stuff in
src/v/datalake/ depend on these new types in translation. Also these
utilities are only used with translation code which makes this move
logical.

bharathv committed:
dl/translation: move exception types to their own header

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/64877-->
<!--66b82435-98c1-494e-8afd-57515f893fd7-->
<!--f5dd588f-71a6-4faa-ad31-916a803dbbf0-->
<!--80bcae32-2cc2-4a0b-aaa5-520a1d79be76-->
<!--title-->
## Retry command for [Build#64877](https://buildkite.com/redpanda/redpanda/builds/64877)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```


/ci-repeat 1
tests/rptest/tests/partition_force_reconfiguration_test.py::PartitionForceReconfigurationTest.test_basic_reconfiguration@{"acks":-1,"controller_snapshots":true,"restart":true}
tests/rptest/tests/datalake/disk_budget_test.py::DatalakeDiskUsageTest.test_idle_finish@{"cloud_storage_type":1,"concurrent_translations":4,"num_partitions":10}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"enable_failures":true,"mixed_versions":false,"with_chunked_compaction":false,"with_iceberg":true,"with_tiered_storage":false}
```
<!--/body-->

bharathv committed:
wip: remove finish_translation_requested logic

secpanda commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devex-c3w/pr-checks/bc5a8294-9b72-42f1-9fab-234e343054a0?refs=53836f84e6e6b8438e0c0bcf79d6d30ca2f3ca25&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devex-c3w/pr-checks/bc5a8294-9b72-42f1-9fab-234e343054a0/license?refs=53836f84e6e6b8438e0c0bcf79d6d30ca2f3ca25&source=prChecksComment)

secpanda commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devprod/pr-checks/992c6dc8-b203-403b-80f7-a64741d7f75a?refs=53836f84e6e6b8438e0c0bcf79d6d30ca2f3ca25&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devprod/pr-checks/992c6dc8-b203-403b-80f7-a64741d7f75a/license?refs=53836f84e6e6b8438e0c0bcf79d6d30ca2f3ca25&source=prChecksComment)

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/65159-->
<!--d42585b4-d71b-491d-8483-6abb5f761f94-->
<!--db316ed7-afb5-4b48-b083-2972e7c5b2d4-->
<!--ae1d4394-8bc9-4f2d-a670-56e250835022-->
<!--title-->
## Retry command for [Build#65159](https://buildkite.com/redpanda/redpanda/builds/65159)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```


/ci-repeat 1
tests/rptest/tests/datalake/translation_interruption_test.py::DatalakeTranslationInterruptionsTest.test_oom_interruptions@{"catalog_type":"rest_hadoop","cloud_storage_type":1,"query_engine":"spark"}
tests/rptest/tests/tiered_storage_model_test.py::TieredStorageTest.test_tiered_storage@{"cloud_storage_type_and_url_style":[1,"path"],"test_case":{"name":"(TS_Read == True, TS_TxRangeMaterialized == True)"}}
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_conflicting_names
tests/rptest/transactions/producers_api_test.py::ProducersAdminAPITest.test_producers_state_api_during_load
tests/rptest/tests/datalake/disk_budget_test.py::DatalakeDiskUsageTest.test_idle_finish@{"cloud_storage_type":1,"concurrent_translations":4,"num_partitions":10}
tests/rptest/tests/simple_e2e_test.py::SimpleEndToEndTest.test_consumer_interruption
```
<!--/body-->
